### PR TITLE
Added email retry logic for failed batches

### PR DESCRIPTION
**Pull request by naz**
ronseal

### CODE REVIEW

File: core/server/services/mega/mega.js [hunk 1]
```diff
@@ -187,7 +202,7 @@ async function listener(emailModel, options) {
     });
 
     let meta = [];
-    let error;
+    let error = null;
```
Comment (gargol) [inline core/server/services/mega/mega.js]: Was initially inclined to set this value in `retryFailedEmail` method, but here it seems more suitable as we already modify `error` property anyways.

File: core/server/services/mega/mega.js [hunk 2]
```diff
@@ -228,14 +243,24 @@ async function listener(emailModel, options) {
     }
 }
 
+const statusChangeListener = (emailModel, options) => {
+    const emailRetried = emailModel.wasChanged() && (emailModel.get('status') === 'pending') && (emailModel.previous('status') === 'failed');
+
+    if (emailRetried) {
+        listener(emailModel, options);
```
Comment (gargol) [inline core/server/services/mega/mega.js]: Keeping this retry logic in a listener to behave in similar async fashion as the main listener does.

File: core/server/services/mega/mega.js [hunk 3]
```diff
@@ -228,14 +243,24 @@ async function listener(emailModel, options) {
...
+    }
+};
+
 function listen() {
     common.events.on('email.added', listener);
+    common.events.on('email.edited', statusChangeListener);
```
Comment (gargol) [inline core/server/services/mega/mega.js]: Don't think it's a good name - `statusChangeListener`. It is meant to indicate that the previously failed emails batch was marked as `pending` again (meaning we can resend it). Also the other method name `listener` is way to generic now. Probably will rename it to `pendingEmailHandler` :thinking:

File: core/server/api/canary/posts.js [hunk 4]
```diff
@@ -149,9 +149,18 @@ module.exports = {
                         return model;
                     }
 
-                    if (!model.get('email') && (model.get('status') === 'published') && model.wasChanged()) {
-                        const email = await mega.addEmail(model.toJSON());
-                        model.set('email', email);
+                    const postPublished = model.wasChanged() && (model.get('status') === 'published') && (model.previous('status') !== 'published');
```
Comment (gargol) [inline core/server/api/canary/posts.js]: The controller logic starts to bloat a bit with email handling stuff. Would need to think up on how to handle it better if more code is added.

**Review by kevinansfield — APPROVED** [review]

**Pull request merged**